src/runtime/entrypoints/main.ts

```diff
@@ -24,49 +32,266 @@ function createRootFragment(
+ * be rendered in a flattened manner in the DOM.
+ *
+ * Example:
+ *   <div>
+ *     <!--frsh-island:0-->
+ *     <!--frsh-slot:children-->
+ *     <p>server content</p>
+ *     <!--/frsh-slot:children-->
+ *     <!--/frsh-island:0-->
+ *   </div>
+ *
+ * Here we have a flat DOM structure, but from the virtual-dom
+ * perspective we should render:
+ *   <div> -> <Island> -> ServerComponent -> <p>server content</p>
+ *
+ * To solve this we're keeping track of the virtual-dom hierarchy
+ * in a stack-like manner, but do the actual iteration in a list-based
+ * fashion over an HTMLElement's children list.
+ */
+function _walkInner(
+  islands: Record<string, ComponentType>,
+  // deno-lint-ignore no-explicit-any
+  props: any[],
+  markerStack: Marker[],
+  vnodeStack: VNode[],
+  node: Node | Comment,
+) {
+  let sib: Node | null = node;
+  while (sib !== null) {
+    const marker = markerStack.length > 0
+      ? markerStack[markerStack.length - 1]
+      : null;
+
+    // We use comment nodes to mark fresh islands and slots
+    if (isCommentNode(sib)) {
+      let comment = sib.data;
+      if (comment.startsWith("!--")) {
+        comment = comment.slice(3, -2);
+      }
+
+      if (comment.startsWith("frsh-slot")) {
+        // Note: Nested slots are not possible as they're flattened
+        // already on the server.
+        markerStack.push({
+          startNode: sib,
+          text: comment,
+          endNode: null,
+          kind: MarkerKind.Slot,
+        });
+        // @ts-ignore TS gets confused
+        vnodeStack.push(h(ServerComponent, { key: comment }));
+      } else if (
+        marker !== null && (
+          comment.startsWith("/frsh") ||
+          // Check for old Preact RTS
+          marker.text === comment
+        )
+      ) {
+        // We're closing either a slot or an island
+        marker.endNode = sib;
+
+        markerStack.pop();
+        const parent = markerStack.length > 0
+          ? markerStack[markerStack.length - 1]
+          : null;
+
+        if (marker.kind === MarkerKind.Slot) {
+          // If we're closing a slot than it's assumed that we're
+          // inside an island
+          if (parent?.kind === MarkerKind.Island) {
+            const vnode = vnodeStack.pop();
+
+            // For now only `props.children` is supported.
+            const islandParent = vnodeStack[vnodeStack.length - 1]!;
+            // Overwrite serialized `{__slot: "children"}` with the
+            // actual vnode child.
+            islandParent.props.children = vnode;
+          }
+
+          // Remove markers
+          marker.startNode.remove();
+          sib = sib.nextSibling;
+          marker.endNode.remove();
+          continue;
+        } else if (marker.kind === MarkerKind.Island) {
+          // We're ready to revive this island if it has
+          // no roots of its own. Otherwise we'll treat it
+          // as a standard component
+          if (markerStack.length === 0) {
```

Comment:
Essentially this means that in this configuration, only the outermost island is explicitly revived, right? The inner islands are revived because it is part of the outer islands' subgraph.

```
// route/index.tsx

<Island n={1}>
  <div>
    <Island n={2}>
      <div>Hello</div>
    </Island>
  </div>
</Island>
```

So that means that for both the "inner" and "outer" island, we need to serialize and re-inject props.

---

This is interesting, because you can already "nest" islands inside of each other in a slightly different way:

```
// routes/index.tsx

<OuterIsland />
```

// islands/OuterIsland.tsx

<div><InnerIsland /></div>

// islands/InnerIsland.tsx
<div>Inner!</div>
```

Here, the `InnerIsland` is really just a regular component child of the original island. However, in the current island serializer logic, we still serialize the props for the inner island (but I don't think it gets revived).

Right now the docs say that this is a totally unsupported configuration (because no tests), but clarifying / fixing this is probably also a good follow up.

Anyway - nothing for this PR :sweat_smile:

Reply:
Ohh that's a good point, that should be doable without much effort to keep serializing props and revive from that. I'll look into that.

Reply:
Nevermind! Turns out this already works as expected in the PR. Reason is that as soon as we encounter the opening marker for an island we instantiate the `vnode` and deserialize the props. This path is hit for every island regardless if it's nested or not. Only once we exit the recursion and hit the end marker we'll overwrite the vnode's children.

Added a test for that though. That was indeed missing. Thanks for pointing that out 👍